### PR TITLE
Report dropped connections.

### PR DESCRIPTION
**Pull request by bosilca**
This is a collective fix for the issue #2320. Instead of silently dropping connections that do not match the IP addresses known for the peer, we nicely inform the user about the mismatch.

### CODE REVIEW

File: opal/util/net.c [hunk 1]
```diff
@@ -405,7 +403,6 @@ opal_net_get_hostname(const struct sockaddr *addr)
        int err = errno;
        opal_output (0, "opal_sockaddr2str failed:%s (return code %i)\n",
                     gai_strerror(err), error);
-       free (name);
```
Comment (jsquyres) [inline opal/util/net.c]: @bosilca Why did you remove these 3 `free()` statements?

Comment (bosilca) [inline opal/util/net.c]: Because I was curious why we release this buffer in all error cases but never on the successful path, so I went read the code of the get_hostname_buffer function. The returned buffer is associated with the calling thread (TLS), and we better don't free it at all.

Comment (jsquyres) [inline opal/util/net.c]: Why is it associated with TLS data?

Comment (bosilca) [inline opal/util/net.c]: To be similar to inet_ntoa I guess (the returned string is in a static area that is overwritten by successive calls).

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3892e955c74781a6f4eb9e7b8e08fcf5

**Comment by bosilca** [comment]
This has been integrated into the master in ec4a235e6a5 and then by #3077.
